src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -291,13 +291,13 @@ public static unsafe SocketError Send(SafeSocketHandle handle, ReadOnlySpan<byte
             return SocketError.Success;
         }
 
-        public static unsafe SocketError SendFile(SafeSocketHandle handle, SafeFileHandle? fileHandle, byte[]? preBuffer, byte[]? postBuffer, TransmitFileOptions flags)
+        public static unsafe SocketError SendFile(SafeSocketHandle handle, SafeFileHandle? fileHandle, ReadOnlySpan<byte> preBuffer, ReadOnlySpan<byte> postBuffer, TransmitFileOptions flags)
```

Comment:
I'm not sure what happens when you pin an empty Span. Do you get null? Do you get a valid pointer to some random empty array that actually gets pinned? If the latter, that seems bad

Seems like we should ensure that we pass null to TransmitFile for empty spans.

Reply:
Just chatted with @stephentoub -- doing (fixed byte* p = someSpan) will indeed give you null if the span is empty.

Would probably be good to assert this too, though.

Reply:
>  pin an empty Span.

Results in a null-pointer / IntPtr.Zero for this usage after the cast. This is safe.

Reply:
Didn't see the new comment before writing mine.

The casts https://github.com/dotnet/runtime/blob/878fdf3a3b8f9f8edc65c1177ea4592424c54dc0/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L299 make it `IntPtr.Zero` and for the length `0` will be passed to `TransmitFileHelper` where these lengths are checked https://github.com/dotnet/runtime/blob/878fdf3a3b8f9f8edc65c1177ea4592424c54dc0/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L1036 so this operation is safe.

I don't understand what should be asserted here.

Reply:
Yeah, you are right, we are checking the length again in TransmitFileHelper so it doesn't matter either way.

Reply:
> doing (fixed byte* p = someSpan) will indeed give you null if the span is empty.

Yup.  Just sharing in the name of completeness, that logic lives here:
https://github.com/dotnet/runtime/blob/8c76312b621e2cdb22ca0ee603c5b0e1919134a2/src/libraries/System.Private.CoreLib/src/System/Span.cs#L255-L262
When you write:
```C#
fixed (byte* p = someSpan)
```
the C# compiler treats that as equivalent to if you wrote:
```C#
fixed (byte* p = &someSpan.GetPinnableReference())
```
thus invoking the previously cited code, which explicitly checks for 0-length and returns null in that case.

Note that this is not the case if you instead used:
```C#
fixed (byte* p = &MemoryMarshal.GetReference(someSpan))
```
whose sole reason to exist is to not perform that 0-length check.